cmake/external_protobuf.cmake

```diff
@@ -0,0 +1,121 @@
+            ext_protobuf
+            PREFIX ext_protobuf
+            GIT_REPOSITORY ${PROTOBUF_GIT_REPO_URL}
+            GIT_TAG ${PROTOBUF_GIT_BRANCH}
+            INSTALL_COMMAND ""
+            UPDATE_COMMAND ""
+            PATCH_COMMAND ""
+            CONFIGURE_COMMAND ./autogen.sh && ./configure --disable-shared CXXFLAGS=-fPIC
+            BUILD_COMMAND ${MAKE}
+            TMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/tmp"
+            STAMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/stamp"
+            DOWNLOAD_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/download"
+            SOURCE_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            BINARY_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf"
+            BUILD_BYPRODUCTS "${EXTERNAL_PROJECTS_ROOT}/protobuf/src/src/.libs/libprotobuf.a"
+            EXCLUDE_FROM_ALL TRUE
+            )
+    else()
+        # To speed things up prefer 'shallow copy' for CMake 3.6 and later
+        ExternalProject_Add(
+            ext_protobuf
+            PREFIX ext_protobuf
+            GIT_REPOSITORY ${PROTOBUF_GIT_REPO_URL}
+            GIT_TAG ${PROTOBUF_GIT_BRANCH}
+            GIT_SHALLOW TRUE
+            INSTALL_COMMAND ""
+            UPDATE_COMMAND ""
+            PATCH_COMMAND ""
+            CONFIGURE_COMMAND ./autogen.sh && ./configure --disable-shared CXXFLAGS=-fPIC
+            BUILD_COMMAND ${MAKE}
+            TMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/tmp"
+            STAMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/stamp"
+            DOWNLOAD_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/download"
+            SOURCE_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            BINARY_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf"
+            BUILD_BYPRODUCTS "${EXTERNAL_PROJECTS_ROOT}/protobuf/src/src/.libs/libprotobuf.a"
+            EXCLUDE_FROM_ALL TRUE
+            )
+    endif()
+endif()
+
+# -----------------------------------------------------------------------------
+
+ExternalProject_Get_Property(ext_protobuf SOURCE_DIR BINARY_DIR)
+
+# -----------------------------------------------------------------------------
+# Use the interface of FindProtobuf.cmake
+# -----------------------------------------------------------------------------
+
+set(Protobuf_SRC_ROOT_FOLDER ${SOURCE_DIR})
+
+set(Protobuf_PROTOC_EXECUTABLE ${Protobuf_SRC_ROOT_FOLDER}/src/protoc)
+set(Protobuf_INCLUDE_DIR ${Protobuf_SRC_ROOT_FOLDER}/src)
+set(Protobuf_LIBRARY ${Protobuf_SRC_ROOT_FOLDER}/src/.libs/libprotobuf.a)
+set(Protobuf_LIBRARIES ${Protobuf_LIBRARY})
+
+if (NOT TARGET protobuf::libprotobuf)
+    add_library(protobuf::libprotobuf UNKNOWN IMPORTED)
+    set_target_properties(protobuf::libprotobuf PROPERTIES
+        INTERFACE_INCLUDE_DIRECTORIES "${Protobuf_INCLUDE_DIR}"
+        IMPORTED_LOCATION "${Protobuf_LIBRARY}")
+    add_dependencies(protobuf::libprotobuf ext_protobuf)
```

Comment:
use target_link_libraries as in external_gtest. You should not need to set Protobuf_INCLUDE_DIR or any of the other things you set on lines 103-106. Then to use the protobuf library all you need is to include protobuf::libprotobuf in your target_link_libraries of the object that wants to use this lib. Look at how gtest was done as an external project.

Reply:
There are two reasons why I did such way:
1. It recreates portions of the FindProtobuf API. Soon we will add libonnx which depends on libprotobuf. The module uses find_package() macro to configure protobuf. If the macro will not detect variables i.e. Protobuf_INCLUDE_DIR and Protobuf_LIBRARY it will not honor the cache and it will pull the Protobuf from the system.
2. The include directory ${Protobuf_INCLUDE_DIR} as well as ${Protobuf_LIBRARIES} does not exist at the build configuration state. Use of target_link_libraries with i.e. protobuf::libprotobuf results in error:
```
CMake Error in src/ngraph/frontend/onnx_import/CMakeLists.txt:
  Imported target "protobuf::libprotobuf" includes non-existent path

    "/home/awojcik/Snapshots/ngraph/build.debug/protobuf/src/src"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```